doc/developer/design/20250515_query_lifecycle_extensions.md

```diff
@@ -0,0 +1,128 @@
+        - `ExecutionBegan` (already exists, emitted in the Coordinator's `begin_statement_execution`)
+        - `OptimizationBegan` (new event, but not mentioned above, because it's orthogonal to the prepared statement issues) (this should be very close in time to `ExecutionBegan` if things are behaving correctly, but I have a suspicion that this is not always the case)
+        - `OptimizationFinished` (already exists)
+        - `StorageDependenciesFinished` (already exists)
+        - `ComputeDependenciesFinished` (already exists)
+        - `ExecutionFinished` (roughly corresponds to first byte of the result sent to user)
+
+For a query submitted through 4.:
+- `FrontendMessage::Parse` arrives in `protocol.rs`, to which
+    - `mz_statement_lifecycle_history` would get the following events:
+        - `StatementReceived`
+        - `ParsingFinished`
+        - `PrepareFinished`
+- `FrontendMessage::Bind` arrives in `protocol.rs`, to which
+    - `mz_statement_lifecycle_history` would get the following events:
+        - `BindReceived`
+        - `BindFinished` (whose timestamp would probably be very close to `BindReceived`, but later we might do more interesting things at bind time, in which case binding will take more time)
+- `FrontendMessage::Execute` arrives in `protocol.rs`, to which
+    - `mz_statement_lifecycle_history` would get the following events:
+        - `ExecuteReceived`
+        - `ExecuteFinished`
+    - `mz_statement_execution_lifecycle_history` would get the following events:
+        - `ExecutionMessageReceived` (with the same timestamp as the above `ExecuteReceived`)
+        - `ExecutionBegan`
+        - `OptimizationBegan`
+        - `OptimizationFinished`
+        - `StorageDependenciesFinished`
+        - `ComputeDependenciesFinished`
+        - `ExecutionFinished` (time will be the same as the above `ExecuteFinished`)
+
+## Minimal Viable Prototype
+
+I have a limited prototype that doesn't add new builtin tables or change schemas in existing ones, but just adds events to the existing `mz_statement_lifecycle_history` without regard to prepared statements:
+https://github.com/MaterializeInc/materialize/pull/32424
+The prototype shows how we can transfer information between `protocol.rs` and the already existing lifecycle logging code in the `Coordinator` through portals.
+
+## Alternatives
+
+One could imagine trying to keep things simple by just quickly implementing the events only for statements that are not prepared statements. however, it turns out that [a big user](https://github.com/MaterializeInc/accounts/issues/3) is actually using prepared statements, because PowerBI always uses the "Extended Query" flow of pgwire. And since having visibility into that user's PowerBI query latency is one of the main motivations of this work, we should do the right thing and support prepared statements.
```

Comment:
Another alternative is to consider replacing the entire statement logging mechanism. The current one has proven to not scale well and is causing incidents from time to time, so we'll have to rethink it eventually. This will be a major piece of work, possibly involving the integration of some external component, so I can understand if you consider this out of scope. But there is a real risk that the query lifecycle will run into the same scalability issues the statement log experiences once we start leaning more on it.